### PR TITLE
Initial implementation

### LINKED ISSUE #1: Implement .NET RFC 0005
## Describe the Enhancement

[.NET Core RFC 0005](https://github.com/paketo-buildpacks/rfcs/blob/ccfbcb20e4643d9883998ae47ae7a37c92aa3f90/text/dotnet-core/0005-simplify-runtime-dependency.md) has been approved and merged.

## Possible Solution

Create a buildpack that meets the following criteria:
- [x] Metadata
  - [x] `name` should be Paketo Buildpack for ASP.NET Core Runtime, per [RFC 0050](https://github.com/paketo-buildpacks/rfcs/blob/7a44e4e335b776917bbf1ce8d23c3baed14a0b50/text/0050-buildpack-rename.md)
  - [x] `homepage` links back to the buildpack repo
  - [x] `license` has the correct license information
  - [x] Dependency URIs are Microsoft download URIs, NOT URIs from the Paketo dep server
- [x] Detection
  - [x] `provide``dotnet-core-aspnet-runtime`
  - [x] By default, `require` nothing
  - [x] If `$BP_DOTNET_FRAMEWORK_VERSION` is set at build time, `require` `dotnet-core-aspnet-runtime` with the requested version
- [x] Build
  - [x] If `*.runtimeconfig.json` does not exist in the working directory, return immediately.
  - [x] If `*.runtimeconfig.json` exists but does not require any .NET runtimes, return immediately.
  - [x] If `*.runtimeconfig.json` exists and lists required .NET runtimes, install the ASP.NET Core dependency as [distributed by Microsoft](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) into a layer that is available at `launch`.  The layer should be added to the launch-time `$PATH` and `$DOTNET_ROOT` should be set to the layer path in the launch environment.

## Motivation

 This is part of https://github.com/paketo-buildpacks/dotnet-core/issues/755.

### PR DESCRIPTION
**Pull request by ForestEckhardt**
Resolves #1 

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: buildpack.toml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    sha256 = "ca3dc696af0a9fc5c7ce052eba38ecf723cbc30d1dc29d8f85c201eff534d73b"
+    source = "https://download.visualstudio.microsoft.com/download/pr/de3f6658-5d5b-4986-aeb1-7efdf5818437/7df572051df15117a0f52be1b79e1823/aspnetcore-runtime-6.0.4-linux-x64.tar.gz"
+    source_sha256 = "ca3dc696af0a9fc5c7ce052eba38ecf723cbc30d1dc29d8f85c201eff534d73b"
+    stacks = ["io.buildpacks.stacks.bionic"]
+    uri = "https://download.visualstudio.microsoft.com/download/pr/de3f6658-5d5b-4986-aeb1-7efdf5818437/7df572051df15117a0f52be1b79e1823/aspnetcore-runtime-6.0.4-linux-x64.tar.gz"
+    version = "6.0.4"
+
+  [[metadata.dependency-constraints]]
+    constraint = "3.1.*"
+    id = "dotnet-aspnetcore"
```
Comment (sophiewigmore) [inline buildpack.toml:34]: I believe the `id` will need to match the dependency ID `"dotnet-core-aspnet-runtime"`

Comment (ForestEckhardt) [reply buildpack.toml]: Yep missed that one should be fixed now.

File: detect.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	return func(context packit.DetectContext) (packit.DetectResult, error) {
+		var requirements []packit.BuildPlanRequirement
+
+		if environment.DotnetFrameworkVersion != "" {
+			requirements = append(requirements, packit.BuildPlanRequirement{
+				Name: "dotnet-core-aspnet-runtime",
+				Metadata: map[string]interface{}{
+					"version-source": "BP_DOTNET_FRAMEWORK_VERSION",
+					"version":        environment.DotnetFrameworkVersion,
+					"launch":         true,
```
Comment (sophiewigmore) [inline detect.go:20]: Why specifically required at launch in this instance?

File: detect.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package dotnetcoreaspnetruntime
+
+import "github.com/paketo-buildpacks/packit/v2"
+
+type Environment struct {
+	DotnetRollForward      string `env:"BP_DOTNET_ROLL_FORWARD"`
```
Comment (sophiewigmore) [inline detect.go:6]: Should this be used anywhere in detection?

Comment (ForestEckhardt) [reply detect.go:6]: What exactly are you referencing?

File: detect.go (hunk 2 above)
Comment (ForestEckhardt) [reply detect.go:20]: I looked at a couple of other buildpacks and we set launch true for the launch bit if the version variable is set so I just went with that idea.

Comment (sophiewigmore) [reply detect.go:20]: I know we'll want it at launch but isn't that the business of other buildpacks?

File: integration/init_test.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	settings.Buildpacks.DotnetCoreRuntime.Online, err = buildpackStore.Get.
+		WithVersion("1.2.3").
+		Execute(root)
+	Expect(err).NotTo(HaveOccurred())
+
+	settings.Buildpacks.DotnetCoreRuntime.Online, err = buildpackStore.Get.
+		WithOfflineDependencies().
+		WithVersion("1.2.3").
+		Execute(root)
+	Expect(err).NotTo(HaveOccurred())
```
Comment (sophiewigmore) [inline integration/init_test.go:72]: Should this be `settings.Buildpacks.DotnetCoreRuntime.Offline`?

File: integration/offline_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+		it("installs the default dotnet aspnet runtime version into a layer", func() {
+			source, err = occam.Source(filepath.Join("testdata", "default"))
+			Expect(err).NotTo(HaveOccurred())
+
+			var logs fmt.Stringer
+			image, logs, err = pack.WithNoColor().Build.
+				WithPullPolicy("never").
+				WithBuildpacks(
+					settings.Buildpacks.DotnetCoreRuntime.Online,
```
Comment (sophiewigmore) [inline integration/offline_test.go:61]: Should be offline buildpack here

File: detect.go (hunk 2 above)
Comment (ForestEckhardt) [reply detect.go:20]: Hmmm just took a look at the buildpacks I thought I saw this behavior in and I am not seeing it. Must be a misremembering I will fix it.

**Comment by sophiewigmore** [comment]
@ForestEckhardt its been a bit so I forgot some context- could you remind me why it's preferable to check the `runtimeconfig.json` for existing and for runtime versions during build, rather than using its existence as detection criteria?

File: detect.go (hunk 3 above)
Comment (sophiewigmore) [reply detect.go:6]: Nevermind, I'm good

**Comment by ForestEckhardt** [comment]
@sophiewigmore We check it at build because this buildpack will now run after `dotnet-publish` and we want to pull the version of the framework from the `runtimeconfig` if at all possible with is generated during publish. Therefore we have to check after the during build.

**Comment by sophiewigmore** [comment]
Ahhh yes, thank you

File: integration/default_test.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+			// check that all required SBOM files are present
+			Expect(filepath.Join(sbomDir, "sbom", "launch", strings.ReplaceAll(buildpackInfo.Buildpack.ID, "/", "_"), "dotnet-core-aspnet-runtime", "sbom.cdx.json")).To(BeARegularFile())
+			Expect(filepath.Join(sbomDir, "sbom", "launch", strings.ReplaceAll(buildpackInfo.Buildpack.ID, "/", "_"), "dotnet-core-aspnet-runtime", "sbom.spdx.json")).To(BeARegularFile())
+			Expect(filepath.Join(sbomDir, "sbom", "launch", strings.ReplaceAll(buildpackInfo.Buildpack.ID, "/", "_"), "dotnet-core-aspnet-runtime", "sbom.syft.json")).To(BeARegularFile())
+
+			// check an SBOM file
+			contents, err = os.ReadFile(filepath.Join(sbomDir, "sbom", "launch", strings.ReplaceAll(buildpackInfo.Buildpack.ID, "/", "_"), "dotnet-core-aspnet-runtime", "sbom.cdx.json"))
+			Expect(err).NotTo(HaveOccurred())
+			Expect(string(contents)).To(ContainSubstring(`"name": ".NET Core ASP.NET Runtime"`))
```
Comment (fg-j) [inline integration/default_test.go:119]:
```suggestion
			Expect(string(contents)).To(ContainSubstring(`"name": "ASP.NET Core Runtime"`))
```

File: integration/layer_reuse_test.go [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+			Expect(err).NotTo(HaveOccurred(), logs.String())
+
+			imageIDs[secondImage.ID] = struct{}{}
+
+			Expect(secondImage.Buildpacks).To(HaveLen(2))
+			Expect(secondImage.Buildpacks[0].Key).To(Equal(buildpackInfo.Buildpack.ID))
+			Expect(secondImage.Buildpacks[0].Layers).To(HaveKey("dotnet-core-aspnet-runtime"))
+
+			Expect(logs).To(ContainLines(
+				"  Resolving .NET Core ASP.NET Runtime version",
```
Comment (fg-j) [inline integration/layer_reuse_test.go:101]:
```suggestion
				"  Resolving ASP.NET Core Runtime version",
```

File: integration/layer_reuse_test.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+				"    Candidate version sources (in priority order):",
+				`      runtimeconfig.json -> "6.0.0"`,
+				`      <unknown>          -> ""`,
+				"",
+				"    No exact version match found; attempting version roll-forward",
+				"",
+				MatchRegexp(`    Selected .NET Core ASP.NET Runtime version \(using runtimeconfig.json\): 6\.0\.\d+`),
```
Comment (fg-j) [inline integration/layer_reuse_test.go:108]:
```suggestion
				MatchRegexp(`    Selected ASP.NET Core Runtime version \(using runtimeconfig.json\): 6\.0\.\d+`),
```

File: integration/layer_reuse_test.go [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+			Expect(err).NotTo(HaveOccurred(), logs.String())
+
+			imageIDs[secondImage.ID] = struct{}{}
+
+			Expect(secondImage.Buildpacks).To(HaveLen(2))
+			Expect(secondImage.Buildpacks[0].Key).To(Equal(buildpackInfo.Buildpack.ID))
+			Expect(secondImage.Buildpacks[0].Layers).To(HaveKey("dotnet-core-aspnet-runtime"))
+
+			Expect(logs).To(ContainLines(
+				"  Resolving .NET Core ASP.NET Runtime version",
```
Comment (fg-j) [inline integration/layer_reuse_test.go:204]:
```suggestion
				"  Resolving ASP.NET Core Runtime version",
```

File: integration/layer_reuse_test.go [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+				"    Candidate version sources (in priority order):",
+				`      BP_DOTNET_FRAMEWORK_VERSION -> "6.*"`,
+				`      runtimeconfig.json          -> "6.0.0"`,
+				`      <unknown>                   -> ""`,
+				"",
+				MatchRegexp(`    Selected .NET Core ASP.NET Runtime version \(using BP_DOTNET_FRAMEWORK_VERSION\): \d+\.\d+\.\d+`),
```
Comment (fg-j) [inline integration/layer_reuse_test.go:210]:
```suggestion
				MatchRegexp(`    Selected ASP.NET Core Runtime version \(using BP_DOTNET_FRAMEWORK_VERSION\): \d+\.\d+\.\d+`),
```

File: build.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+					Name: "dotnet-core-aspnet-runtime",
+					Metadata: map[string]interface{}{
+						"version":        frameworkVersion,
+						"version-source": "runtimeconfig.json",
+					},
+				},
+			}, context.Plan.Entries...)
+		}
+
+		logger.Process("Resolving .NET Core ASP.NET Runtime version")
```
Comment (fg-j) [inline build.go:85]:
```suggestion
		logger.Process("Resolving ASP.NET Core Runtime version")
```

File: build.go [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+		logger.Process("Executing build process")
+
+		dotnetCoreAspnetRuntimeLayer, err = dotnetCoreAspnetRuntimeLayer.Reset()
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		dotnetCoreAspnetRuntimeLayer.Launch, dotnetCoreAspnetRuntimeLayer.Build, dotnetCoreAspnetRuntimeLayer.Cache = launch, build, build
+
+		logger.Subprocess("Installing .NET Core ASP.NET Runtime %s", dependency.Version)
```
Comment (fg-j) [inline build.go:144]:
```suggestion
		logger.Subprocess("Installing ASP.NET Core Runtime %s", dependency.Version)
```

File: runtime_config_parser.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	buffer := bytes.NewBuffer(nil)
+	err = jsmin.Min(file, buffer)
+	if err != nil {
+		return "", err
+	}
```
Comment (fg-j) [inline runtime_config_parser.go:57]: What's the rationale for minifying the JSON before decoding it?

File: build.go [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+				},
+			}, context.Plan.Entries...)
+		}
+
+		logger.Process("Resolving .NET Core ASP.NET Runtime version")
+
+		priorities := []interface{}{
+			"BP_DOTNET_FRAMEWORK_VERSION",
+			regexp.MustCompile(`.*\.(cs)|(fs)|(vb)proj`),
+			"runtimeconfig.json",
```
Comment (fg-j) [inline build.go:90]:
```suggestion
			"BP_DOTNET_FRAMEWORK_VERSION",
			"runtimeconfig.json",
			regexp.MustCompile(`.*\.(cs)|(fs)|(vb)proj`),
```
I think the version extracted from the `runtimeconfig.json` should take precedence over the one from the projfile. Since the version in the `runtimeconfig.json` is our best and most up-to-date guess about what runtime version the app needs.

File: build.go [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+			if errors.Is(err, os.ErrNotExist) {
+				logger.Process("Skipping build process as there was no runtimeconfig.json found")
+				logger.Break()
+				return packit.BuildResult{}, nil
+			}
+			return packit.BuildResult{}, err
+		}
+
+		if frameworkVersion == "" {
+			logger.Process("Skipping build process as there were no frameworks specified in the runtimeconfig.json")
```
Comment (fg-j) [inline build.go:70]: Would it be more clear for users if the logs said something like "Skipping build process because we think the app is self-contained"?

File: integration/layer_reuse_test.go [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+
+			Eventually(func() string {
+				cLogs, err := docker.Container.Logs.Execute(secondContainer.ID)
+				Expect(err).NotTo(HaveOccurred())
+				return cLogs.String()
+			}).Should(ContainSubstring(".NET runtimes installed"))
+
+			Expect(secondImage.Buildpacks[0].Layers["dotnet-core-aspnet-runtime"].SHA).To(Equal(firstImage.Buildpacks[0].Layers["dotnet-core-aspnet-runtime"].SHA))
+		})
+	})
```
Comment (fg-j) [inline integration/layer_reuse_test.go:129]: I think this is an opportunity to use the new `BuildpackForKey` helper
https://github.com/paketo-buildpacks/occam/pull/165

**Review by fg-j — CHANGES_REQUESTED** [review]
Mostly nitpicks, but a couple of key questions:
1) the name of the buildpack
2) the priority order of version resolution

File: build.go (hunk 15 above)
Comment (sophiewigmore) [reply build.go:70]: I like that it adds the detail that it did the check on the runtimeconfig.json. How about, `Skipping build process, app detected as self-contained due to no frameworks in the runtimeconfig.json` ? Too wordy?

File: .github/.syncignore [hunk 17]
```diff
@@ -0,0 +1 @@
+CODEOWNERS
```
[the remark this answers is not in the document]
Comment (sophiewigmore) [reply .github/.syncignore:1]: Do we need `workflows/update-dependencies.yml` too? And to remove that workflow from here since it won't work?

File: runtime_config_parser.go (hunk 13 above)
Comment (ForestEckhardt) [reply runtime_config_parser.go:57]: This is to remove comments inside of runtimeconfig.json (which is a thing that .NET allows)

Comment (fg-j) [reply runtime_config_parser.go:57]: 🤦 oh, right. thanks for the reminder.

File: integration/layer_reuse_test.go [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+				).
+				Execute(name, source)
+			Expect(err).NotTo(HaveOccurred(), logs.String())
+
+			imageIDs[secondImage.ID] = struct{}{}
+
+			Expect(secondImage.Buildpacks).To(HaveLen(2))
+			secondImageBuildpackMetadata, err := secondImage.BuildpackForKey(buildpackInfo.Buildpack.ID)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(secondImage.Buildpacks[0].Layers).To(HaveKey("dotnet-core-aspnet-runtime"))
```
Comment (fg-j) [inline integration/layer_reuse_test.go:100]:
```suggestion
			Expect(secondImageBuildpackMetadata.Layers).To(HaveKey("dotnet-core-aspnet-runtime"))
```

File: buildpack.toml [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+  pre-package = "./scripts/build.sh"
+
+  [[metadata.dependencies]]
+    cpe = "cpe:2.3:a:microsoft:asp.net_core:3.1:*:*:*:*:*:*:*"
+    deprecation_date = "2024-11-08T00:00:00Z"
+    id = "dotnet-core-aspnet-runtime"
+    name = "ASP.NET Core Runtime"
+    licenses = ["MIT", "MIT-0"]
+    purl = "pkg:generic/dotnet-aspnetcore@3.1.28?checksum=a3fe5848d88dc67cbfceaf0b8e6d3e1178f8441bfedbb21861b2e72fdee3f28d&download_url=https://download.visualstudio.microsoft.com/download/pr/effaa5bf-0fa7-4e5a-9ce8-9ac04ee86669/5afb2b1c2ad68550cec914d8fb303d20/aspnetcore-runtime-3.1.28-linux-x64.tar.gz"
+    sha256 = "a3fe5848d88dc67cbfceaf0b8e6d3e1178f8441bfedbb21861b2e72fdee3f28d"
```
Comment (fg-j) [inline buildpack.toml:25]: should we switch to using `checksum` + the published SHA512 now that it's possible in jam and packit?

Comment (fg-j) [reply buildpack.toml:25]: or do you want to defer that to a subsequent PR?

Comment (ForestEckhardt) [reply buildpack.toml:25]: I can take a quick crack at it.

**Review by fg-j — CHANGES_REQUESTED** [review]
@ForestEckhardt I just experimented with combining this buildpack with the existing rearchitecture spike, and I'm seeing the running app fail to locate the ASP.NET Core runtime, even though the runtime is present in a layer on the app image. The app fails to start with:
```
docker run -it --rm --env PORT=8080 test-rearch
Setting ASPNETCORE_URLS=http://0.0.0.0:8080
You must install .NET to run this application.

App: /workspace/source-app
Architecture: x64
App host version: 6.0.9
.NET location: Not found

Learn about runtime installation:
https://aka.ms/dotnet/app-launch-failed

Download the .NET runtime:
https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=x64&rid=ubuntu.18.04-x64&apphost_version=6.0.9
```

However, running 
```
DOTNET_ROOT=/layers/paketo-buildpacks_dotnet-core-aspnet-runtime/dotnet-core-aspnet-runtime /workspace/source-app
```

inside the container makes the app start as expected. I believe this buildpack must set `DOTNET_ROOT`.

**Comment by ForestEckhardt** [comment]
@fg-j After our short investigation is appears that the `DOTNET_ROOT` needs to be set for FDEs I will make a commit shortly.

**Comment by sophiewigmore** [comment]
@ForestEckhardt @fg-j I'm surprised that `DOTNET_ROOT` needs to be set if the runtime is already on the `PATH`. Could you give a bit more context about why that is Forest?

**Comment by ForestEckhardt** [comment]
@sophiewigmore It have to do with how FDE works. Because we are not invoking the `dotnet` cli the FDE is like trying to look in a default install location which we have not installed the installation in. Because of that the warning message given tells us we need to specify the `DOTNET_ROOT` if we are in the non-default location.

However because of our restructure we can get away with using the `default` directive instead of the `override` directive which is the preferred option.

**Comment by sophiewigmore** [comment]
@ForestEckhardt thank you! that makes sense

**Comment by sophiewigmore** [comment]
@fg-j are you able to re-test with the changes here since you have the re-architecture spike pulled already? Assuming that works now, this PR looks good to me

**Review by fg-j — APPROVED** [review]
I've verified that with the new changes setting `DOTNET_ROOT`, the buildpack now integrates as expected with the other language family buildpacks. I've built and run source apps, FDDs, FDEs, and self-contained apps successfully.
